test/opt/test_passes.cpp

```diff
@@ -0,0 +1,195 @@
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#include <algorithm>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include "opt/libspirv.hpp"
+#include "opt/pass_manager.h"
+#include "opt/passes.h"
+
+namespace {
+
+using namespace spvtools;
+
+// Concatenates a vector of strings into one string. Each string is postfixed
+// with the given |delimiter|.
+std::string Join(const std::vector<std::string>& insts, char delimiter = '\n') {
+  std::ostringstream oss;
+  for (const auto& inst : insts) oss << inst << delimiter;
+  return oss.str();
+}
+
+// Concatenates a vector of strings into one string. Each string is postfixed
+// with the given |delimiter|. If a string contains opcode for debug
+// instruction, that string will be ignored.
+std::string SkipDbgInstJoin(const std::vector<std::string>& insts,
+                            char delimiter = '\n') {
+  // Well, another place requiring the knowledge of the grammar and can be stale
+  // when SPIR-V is updated. It would be nice to automatically generate this,
+  // but the cost is just too high.
+  const std::vector<std::string> dbg_opcodes{
+      // clang-format off
+    "OpSourceContinued", "OpSource", "OpSourceExtension",
+    "OpName", "OpMemberName", "OpString",
+    "OpLine", "OpNoLine", "OpModuleProcessed"
+      // clang-format on
+  };
+
+  std::ostringstream oss;
+  for (const auto& inst : insts) {
+    if (std::all_of(dbg_opcodes.cbegin(), dbg_opcodes.cend(),
+                    [&inst](const std::string& op) {
+                      return inst.find(op) == std::string::npos;
+                    })) {
+      oss << inst << delimiter;
+    }
+  }
+  return oss.str();
+}
+
+// Template class for testing passes. It contains some handy utility methods for
+// running passes and checking results.
+//
+// To write value-Parameterized tests:
+//   using ValueParamTest = PassTest<::testing::TestWithParam<std::string>>;
+// To use as normal fixture:
+//   using FixtureTest = PassTest<::testing::Test>;
+template <typename GT>
```

Comment:
Was to mean **G**oogle**T**est. But, yeah, maybe only I can understand it. Renamed.